### DISCUSSION TITLE
Third parameter of `set` is not included in type definition

### DISCUSSION THREAD

**Discussion by ncksllvn** [pmndrs/zustand#2734] [Bug report] [answered]
### Bug Description

In the snippet below, the `set` function produces the TS error `Expected 1-2 arguments, but got 3.`. It seems that the third parameter, `{ type: 'set/testing' }`, is not included in the type def. 

If you just ignore the TS error and run the code anyway, the code _does_ work and produces the expected result in the Redux dev tools (an action of type `set/testing`).

```ts
type TestState = {
  testing: boolean;
  setTesting: (value: boolean) => void;
}

export const createTestState: StateCreator<TestState, [], [], TestState> = (
  set,
  _get,
  _store
) => ({
  testing: false,
  setTesting(value: boolean) {
    set({ testing: value }, undefined, { type: 'set/testing' })
  }
});

export const createUserStore = () =>
  createStore<TestState>()(
    devtools((set, get, store) => createTestState(set, get, store), {
      name: 'Testing'
    })
  );
```

### Workaround
As a workaround, we did this horrible thing below, to cast `set` into what we need it to be in order to include the third parameter.

```ts
export const createTestState: StateCreator<TestState, [], [], TestState> = (
  _set,
  _get,
  _store
) => {
  const set = _set as ((...params: [...Parameters<typeof _set>, { type: unknown }?]) => ReturnType<typeof _set>);
  return {
    testing: false,
    setTesting(value: boolean) {
      set({ testing: value }, undefined, { type: 'set/testing' })
    }
  }
}
```

### Reproduction Link

https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgYygUwIYzQYQgO2wA94BfOAMyghDgHJ0NkY6BuAKFElkRUewDKMaGgA0cIVlz9occlRr0AXgFcAzjAz4AJm07ho8JNrQA3YRAA2auZWq06qjVu0B6EMG3bLaAO4Z0PXYYAE8wNDgAFTQNSWw4AF5Edjg4bA1gfABzAC44ACMIK0x8DlS1NBhojOy8gApTDEsVNDzC4q0ASkSAPjhTCE8OUnZ2NCJueGQCDT5MbGqYONaJTWwcGSgAHkXl8QBtAF0D46iYpbW0PqS6lLgKmFE7gH0syqfU540Rdm6EvrqCDu6RgmVylCaFQ+90qizBDSaLTaRR8XWSqXKlUBaXOYLyjWaEVI4hUOjQFEyaG04iQoXCeToD1cILBdDknTuI1InQ4YwmhhQMym-DQAFUKlAhCJEnA6n8endUPM0FL0DtzssenLbhi4CZzCi1HU6g9xG9HvdhOh5XMpLtLib3nBzeJvtaaXddfgMCAVnQ4dk6NDUty7jygA

**Comment by dbritto-dev** [comment, marked as answer]
@ncksllvn here you go ->

instead of this

```ts
StateCreator<TestState, [], [], TestState>
```

should be this

```ts
StateCreator<TestState, [['zustand/devtools', never]], [], TestState>
```

**Comment by ncksllvn** [comment]
Wow sorry I missed that in the docs. Thanks @dbritto-dev !
